src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/Shake128.cs

```diff
@@ -87,10 +94,13 @@ public byte[] GetHashAndReset(int outputLength)
             ArgumentOutOfRangeException.ThrowIfNegative(outputLength);
             CheckDisposed();
 
-            byte[] output = new byte[outputLength];
-            _hashProvider.Finalize(output);
-            _hashProvider.Reset();
-            return output;
+            return ConcurrentGuard<int, byte[]>(outputLength, static (int outputLength, ref LiteXof hashProvider) =>
```

Comment:
I think this is at this layer because you don't want LiteXof (or LiteHash) to take the hit.

What if we wrapped LiteXof in HeavyXof and HeavyXof does this before calling the Lite version?  That feels like it'll keep the implementation much more contained... but maybe there's a reason why it's bad.

Reply:
>  LiteXof (or LiteHash) to take the hit

Correct. The lites should stay light, and are used by the streaming one-shots, where the guards won't do anything useful. Streams can also do a lot of appends, so that's a lot of Interlocked Increment/Decrement for nothing.

> What if we wrapped LiteXof in HeavyXof

Yeah. Shake128/256, in my brain, _is_ HeavyXof. I was trying to avoid introducing another layer of indirection and the penalties that come with it. But, it would allow more sharing between Shake128 and Shake256.

I'll play around with the idea and see if I can produce something that doesn't annoy me.